### PR TITLE
[GSoC] Coroutines migration RebuildCram

### PR DESCRIPTION
**Pull request by divyansh-dxn**
Third commit would require more attention. Tried different approaches to tackle this one but one here I found most suitable and understandable. I have kept the callbacks as it is, used the provided callbacks in the suspend function.

Have manually tested the dialog as well. Wasn't able to cover every scenario but yeah, which I tested work. I'd ask to once more test it on real device.

## Fixes
Fixes a part of #7108

## How Has This Been Tested?
Pixel 4a API 30 and Unit tests

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by divyansh-dxn** [comment]
Third commit would require more attention. Tried different approaches to tackle this one but one here I found most suitable and understandable. I have kept the callbacks as it is, used the provided callbacks in the suspend function.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 1]
```diff
@@ -2446,31 +2445,19 @@ open class DeckPicker :
...
-            context.updateDeckList()
-            if (context.fragmented) {
-                context.loadStudyOptionsFragment(false)
+    suspend fun rebuildFiltered(did: DeckId) {
+        withProgress(resources.getString(R.string.rebuild_filtered_deck)) {
+            withCol {
+                decks.select(did)
+                Timber.d("rebuildFiltered: doInBackground - RebuildCram")
+                col.sched.rebuildDyn(col.decks.selected())
+                updateValuesFromDeck(col, true)
```
Comment (ShridharGoel) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2454]: How was this being done earlier? Also, why using `col` explicitly when you are already inside `withCol`?

Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:2454]: Actually it's not an issue `col` inside withCol{} actually picks up `this` instance
```kotlin
    /**
     * Allows a collection to be used as a CollectionGetter
     * @return Itself.
     */
    override val col: Collection
        get() = this
```
but yeah, a cleanup I can do which I forgot.

Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt]: Done !

**Review by Arthur-Milchior — APPROVED** [review]

**Review by ShridharGoel — APPROVED** [review]

**Pull request merged**